prometheus_hardware_exporter/collectors/redfish.py

```diff
@@ -38,28 +56,61 @@ def get_sensor_data(self, host: str, username: str, password: str) -> Dict[str,
             output[name] = sensors
         return output
 
-    def _get_sensor_data(self, host: str, username: str, password: str) -> Optional[List[Any]]:
-        """Return sensor if sensor exists else None."""
+    def _get_sensor_data(
+        self, login_info: dict, timeout: int, max_retry: int
+    ) -> Optional[List[Any]]:
+        """Return sensor if sensor exists else None.
+
+        Params:
+            login_info: a dictionary with host, username and password for redfish
+            timeout: redfish client timeout
+            max_retry: redfish client max retry
+        """
         sensors: Optional[List[Any]] = None
         redfish_obj: Optional[HttpClient] = None
+        host: str = login_info["host"]
+        username: str = login_info["username"]
+        password: str = login_info["password"]
         try:
             redfish_obj = redfish.redfish_client(
-                base_url=host, username=username, password=password
+                base_url=host,
+                username=username,
+                password=password,
+                timeout=timeout,
+                max_retry=max_retry,
             )
             redfish_obj.login(auth="session")
             sensors = redfish_utilities.get_sensors(redfish_obj)
             return sensors
-        except (InvalidCredentialsError, SessionCreationError, ConnectionError) as err:
+        except (
+            InvalidCredentialsError,
+            SessionCreationError,
+            ConnectionError,
+            RetriesExhaustedError,
+        ) as err:
             logger.exception(err)
         finally:
             # Log out
             if redfish_obj:
                 redfish_obj.logout()
         return sensors
 
-    def discover(self) -> bool:
-        """Return true if redfish is been discovered."""
+
+def get_discover(ttl: int) -> Callable:
+    """Return the cached discover function.
+
+    Passes the ttl value to the cache decorator at runtime.
+    """
+
+    @ttl_cache(ttl=ttl)
+    def _discover() -> bool:
+        """Return true if redfish services have been discovered."""
+        logger.info("Discovering redfish services...")
         services = redfish.discover_ssdp()
         if len(services) == 0:
+            logger.info("No redfish services discovered")
             return False
+        logger.info("Discovered redfish services: %s", services)
         return True
+
+    return _discover
```

Comment:
Why we moved this out of the class? Why not:
```python
def discover(self) -> bool:
    @ttl_cache(ttl=self.ttl)
    def _discover() -> bool:
        ...

    return _discover()
```

I do not understand why we are always creating class object, which ware basically storing no values. before you add `__init__` there was no class variable, no variable defined in __init__ and no properties.

Reply:
The example somehow make the `@ttl_cache` not working. The purpose here is just let the decorator can get the config value as the ttl settings.

Reply:
I think we can easily do:
```python
@ttl_cache(ttl=TTF)
def discover(self) -> bool:
...
```
or completely remove this class object and define just simple functions outside of it.

My issue is that we define class and them moved function, which is 100% related with the class, outside of class object only to override bad design of cachetools and our. I do not think that this is blocking, but it's just bad design.

Reply:
Thanks for the comment @rgildein !

The main aim here is to configure the ttl value for the cache at runtime which is coming from the the CLI args defined in `__main__.py`. I initially defined a class attribute which could then be modified at runtime to change the ttl value. But that didn't work since the decorator caches the initial value of the ttl class attribute when the class is initialized and doesn't change even when the attribute it modified later.

After some help from @jneo8 I defined a sort of wrapper function that returns the cached discover method with the ttl value being passed at runtime. 

I don't know why I didn't try this out earlier, but having the method inside the class itself seems to work while testing on a machine and it passes the unit tests as well. So we don't have to have the bad design after all :)

```python
class RedfishHelper:
    """Helper function for redfish."""

    def __init__(self, discover_cache_ttl: int) -> None:
        """Initialize disover method with TTL value."""
        self.discover = self.get_discover(discover_cache_ttl)

    (...)

    def get_discover(self, ttl: int) -> Callable:
        """Return the cached discover function.

        Passes the ttl value to the cache decorator at runtime.
        """

        @ttl_cache(ttl=ttl)
        def _discover() -> bool:
            """Return true if redfish services have been discovered."""
            logger.info("Discovering redfish services...")
            services = redfish.discover_ssdp()
            if len(services) == 0:
                logger.info("No redfish services discovered")
                return False
            logger.info("Discovered redfish services: %s", services)
            return True

        return _discover

Reply:
Like this we are doing wrapper for wrapper, which is actually also wrapper. That is why I was trying to avoid using `get_discover`. But if there is no other solution, it's ok.

Reply:
Yeah I really couldn't think of a more elegant solution to this problem either :disappointed: